src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregatorStore.cs

```diff
@@ -0,0 +1,472 @@
+            where TStringSequence : IStringSequence, IEquatable<TStringSequence>
+            where TObjectSequence : IObjectSequence, IEquatable<TObjectSequence>
+        {
+            TStringSequence? seq = default;
+            if (seq is StringSequence1)
+            {
+                if (Label1 == null)
+                {
+                    Interlocked.CompareExchange(ref Label1, new FixedSizeLabelNameDictionary<StringSequence1, ObjectSequence1, TAggregator>(), null);
+                }
+                return (FixedSizeLabelNameDictionary<TStringSequence, TObjectSequence, TAggregator>)(object)Label1;
+            }
+            else if (seq is StringSequence2)
+            {
+                if (Label2 == null)
+                {
+                    Interlocked.CompareExchange(ref Label2, new FixedSizeLabelNameDictionary<StringSequence2, ObjectSequence2, TAggregator>(), null);
+                }
+                return (FixedSizeLabelNameDictionary<TStringSequence, TObjectSequence, TAggregator>)(object)Label2;
+            }
+            else if (seq is StringSequence3)
+            {
+                if (Label3 == null)
+                {
+                    Interlocked.CompareExchange(ref Label3, new FixedSizeLabelNameDictionary<StringSequence3, ObjectSequence3, TAggregator>(), null);
+                }
+                return (FixedSizeLabelNameDictionary<TStringSequence, TObjectSequence, TAggregator>)(object)Label3;
+            }
+            else if (seq is StringSequenceMany)
+            {
+                if (LabelMany == null)
+                {
+                    Interlocked.CompareExchange(ref LabelMany, new FixedSizeLabelNameDictionary<StringSequenceMany, ObjectSequenceMany, TAggregator>(), null);
+                }
+                return (FixedSizeLabelNameDictionary<TStringSequence, TObjectSequence, TAggregator>)(object)LabelMany;
+            }
+            // we should never get here unless this library has a bug
+            throw new ArgumentException("Unexpected sequence type");
+        }
+
+        public void Collect(Action<LabeledAggregationStatistics> visitFunc)
+        {
+            if (NoLabelAggregator != null)
+            {
+                IAggregationStatistics stats = NoLabelAggregator.Collect();
+                visitFunc(new LabeledAggregationStatistics(stats));
+            }
+            Label1?.Collect(visitFunc);
+            Label2?.Collect(visitFunc);
+            Label3?.Collect(visitFunc);
+            LabelMany?.Collect(visitFunc);
+        }
+    }
+
+    internal struct LabelInstruction
+    {
+        public LabelInstruction(int sourceIndex, string labelName)
+        {
+            SourceIndex = sourceIndex;
+            LabelName = labelName;
+        }
+        public readonly int SourceIndex;
+        public readonly string LabelName;
```

Comment:
Elsewhere with similar structs you had these as auto-props.  As this is all internal, it doesn't really matter which you use, but it'd be nice to be consistent.